pkg/docker/storage.jsx

```diff
@@ -0,0 +1,440 @@
+            return model_promise;
+
+        var dfd = $.Deferred();
+        init_model(function (model) {
+            dfd.resolve(model);
+        });
+
+        model_promise = dfd.promise();
+        return model_promise;
+    }
+
+    var DriveBox = React.createClass({
+        getInitialState: function () {
+            return {
+                drives: this.props.model.extra_devices,
+                checked: { }
+            };
+        },
+        componentDidMount: function () {
+            var self = this;
+            $(this.props.model).on("changed", function () {
+                self.setState({ drives: self.props.model.extra_devices });
+            });
+        },
+
+        driveChecked: function (drive) {
+            return !!this.state.checked[drive.path];
+        },
+        makeDriveToggler: function (drive) {
+            var self = this;
+            return function (event) {
+                self.state.checked[drive.path] = !self.state.checked[drive.path];
+                $(event.currentTarget).find("input").prop("checked", self.state.checked[drive.path]);
+                self.setState({ checked: self.state.checked });
+            };
+        },
+
+        onButtonClicked: function () {
+            var self = this;
+            if (self.props.callback) {
+                var drives = [ ];
+                for (var d in self.state.checked) {
+                    if (self.state.checked[d]) {
+                        for (var i = 0; i < self.state.drives.length; i++) {
+                            if (self.state.drives[i].path == d) {
+                                drives.push(self.state.drives[i]);
+                            }
+                        }
+                    }
+                }
+                self.setState({ checked: { } });
+                self.props.callback(drives);
+            }
+        },
+
+        render: function() {
+            var self = this;
+            var i, d;
+
+            var button_enabled = false;
+            var drive_rows = [ ];
+            var drive_paths = { };
+
+            for (i = 0; i < self.state.drives.length; i++) {
+                var drive = self.state.drives[i];
+
+                if (self.state.checked[drive.path])
+                    button_enabled = true;
+
+                drive_paths[drive.path] = true;
+
+                drive_rows.push(
+                    <tr onClick={self.makeDriveToggler(drive)}>
+                        <td><input type="checkbox"
+                                   checked={self.driveChecked(drive)}/>
+                        </td>
+                        <td><img src="images/drive-harddisk-symbolic.svg"/></td>
+                        <td><span>{drive.name}</span><br/><span>{cockpit.format_bytes(drive.size)} {drive.class}</span></td>
+                    </tr>);
+            }
+
+            for (d in self.state.checked) {
+                if (!drive_paths[d])
+                    self.state.checked[d] = false;
```

Comment:
I think this isn't actually necessary.  The drives are unticked when the "Add" dialog opens, which should be enough.  If a checked drive disappears and then comes back, we might as well keep it checked.